src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -163,6 +169,10 @@ public function normalize($data, string $format = null, array $context = [])
                 return $data;
             }
 
+            if (($context[self::EMPTY_ITERABLE_AS_OBJECT] ?? false) === true && is_countable($data) && 0 === \count($data)) {
```

Comment:
EMPTY_ITERABLE_AS_OBJECT currently absorbs PRESERVE_EMPTY_OBJECTS functionality
![image](https://user-images.githubusercontent.com/11708238/127333964-d2efcf29-93f4-4fdf-8054-229b53396e5c.png)
It is less flexible for me. It might be better to apply new option for plain arrays only

Reply:
I don't think so. See the tests

Reply:
Yes, 2nd test case is the same except plain arrays. So for countable objects new option behaves the same.
```
class Foo
{
    public UserList $list; // implements countable interface
    
    public array $settings; // Array setting => value
}
```
If you want to serialize empty settings as JSON object you will use new option, but at the same time your empty `UserList` will be encoded as object. 
So it's less controllable now.

Reply:
What do you propose?
But anyway, you can put the option only where you need it and not globally.

Reply:
> But anyway, you can put the option only where you need it and not globally.

But as i mentioned before it propagates deeper.


> What do you propose?

```
if (is_countable($data) && \count($data) === 0) {
    switch (true) {
        case ($context[AbstractObjectNormalizer::PRESERVE_EMPTY_OBJECTS] ?? false) && is_object($data):
            return $data;
        case ($context[AbstractObjectNormalizer::EMPTY_ARRAY_AS_OBJECT] ?? false) && is_array($data):
            return new \ArrayObject();
    }
}
```
In this case you can save your current `testNormalizeEmptyIterableAsObject` test case for PRESERVE_EMPTY_OBJECTS + EMPTY_ARRAY_AS_OBJECT. 
So the test cases 

1. PRESERVE_EMPTY_OBJECTS
2. EMPTY_ARRAY_AS_OBJECT
3. PRESERVE_EMPTY_OBJECTS + EMPTY_ARRAY_AS_OBJECT

Reply:
Thanks, I'm gonna to test it asap

Reply:
Thanks for taking time to help with this issue.

I have pushed another implementation.

I also added more tests. And I also run the current tests suite with the old implementation (even before my previous) PR and tests pass (exception the one that use `EMPTY_ARRAYS_AS_OBJECT` obviously).

I hope it's now fine for your :)